pkg/service/deploy.go

```diff
@@ -23,83 +27,104 @@ type DeployService struct {
+
+	if err != nil {
+		return errors.Wrap(err, "error creating local backend")
+	}
 
 	log.Infoln("Creating DCE infrastructure")
-	artifactsBucket := s.createDceInfra(stateBucket, overrides)
+	artifactsBucket, err := s.createDceInfra(ctx, overrides)
+
+	if err != nil {
+		return errors.Wrap(err, "error creating infrastructure")
+	}
+
 	log.Infoln("Artifacts bucket = ", artifactsBucket)
 
 	log.Infoln("Deploying code assets to DCE infrastructure")
 	s.deployCodeAssets(artifactsBucket, overrides)
+	return nil
 }
 
-func (s *DeployService) createRemoteStateBackend(overrides *DeployOverrides) string {
-	tmpDir, originDir := s.Util.MvToTempDir("dce-")
-	defer s.Util.RemoveAll(tmpDir)
+// createTFMainFile creates the main.tf file. The default behavior, without
+// using a local repository, is to create the file with the bare minimum
+// required to use Terraform with local state.
+func (s *DeployService) createTFMainFile(overrides *DeployOverrides, overwrite bool) (string, error) {
+	_, originDir := s.Util.ChToConfigDir()
 	defer s.Util.Chdir(originDir)
 
-	fileName := filepath.Join(tmpDir, "init.tf")
-
-	remoteStateFile := s.getRemoteStateFile()
-	s.Util.WriteFile(fileName, remoteStateFile)
-
-	log.Infoln("Initializing terraform working directory and building remote state infrastructure")
-	s.Util.Init([]string{})
-
-	args := []string{"namespace=" + overrides.Namespace}
-	s.Util.Terraformer.Apply(args)
+	fileName := s.Util.GetLocalBackendFile()
 
-	log.Infoln("Retrieving remote state bucket name from terraform outputs")
-	stateBucket := s.Util.Terraformer.GetOutput("bucket")
-	log.Infoln("Remote state bucket = ", stateBucket)
-
-	return stateBucket
+	if s.Util.IsExistingFile(fileName) && !overwrite {
+		log.Warnln("'main.tf' already exists and overwrite not specified; using existing file")
+	} else {
+		tfMainContents, err := s.getLocalTFMainContents(overrides)
+		if err != nil {
+			return "", err
+		}
+		s.Util.WriteFile(fileName, tfMainContents)
+	}
+	return fileName, nil
 }
 
-func (s *DeployService) createDceInfra(stateBucket string, overrides *DeployOverrides) string {
-	tmpDir, originDir := s.Util.MvToTempDir("dce-")
-	defer s.Util.RemoveAll(tmpDir)
+func (s *DeployService) createDceInfra(ctx context.Context, overrides *DeployOverrides) (string, error) {
+	_, originDir := s.Util.ChToConfigDir()
 	defer s.Util.Chdir(originDir)
 
-	tfModulesDir := s.retrieveTFModules()
-	files := s.Util.ReadDir(tfModulesDir)
+	s.retrieveTFModules()
 
-	s.Util.Chdir(files[0].Name())
+	deployLogFileName := s.Util.GetLogFile()
+	ctx = context.WithValue(ctx, "deployLogFile", deployLogFileName)
 
 	log.Infoln("Initializing terraform working directory")
-	s.Util.Terraformer.Init([]string{"-backend-config=bucket=" + stateBucket, "-backend-config=key=local-tf-state"})
+	s.Util.Terraformer.Init(ctx, []string{})
 
 	log.Infoln("Applying DCE infrastructure")
-
-	args := argsFromOverrides(overrides)
-	s.Util.Terraformer.Apply(args)
+	s.Util.Terraformer.Apply(ctx, []string{})
```

Comment:
Maybe we should remove or change this log statement. The output makes it look like it was already applied before asking if you want to apply dce resources

```
~ ./dce-cli system deploy
Creating DCE infrastructure
Initializing terraform working directory
Applying DCE infrastructure
Are you sure you would like to create DCE resources?
```

Reply:
Nice catch. I'll clean these up to show a better flow.

Reply:
> @nathanagood -- do you think we could do an in person review of this PR, given its size?

@eschwartz - If by "in-person" you mean over vid, sure. I'd hate to wait until we are physically in-person.

Reply:
Updated in latest push. The output now looks like this:

```bash
Initializing
Creating DCE infrastructure
Are you sure you would like to create DCE resources? (must type "yes" if yes)	yes
Retrieving artifacts location
Artifacts bucket =  574035908831-dce-artifacts-dce-bmddqqcb
Deploying code assets to DCE infrastructure
Downloading DCE code assets
.
.
.
.
.
.
.
.
.
.
.
Updating AWS Lambda functions...
Finished updating AWS Lambda functions.
```